### ISSUE TITLE
✨ Reworking the "combobox" component

### ISSUE THREAD

**Issue by substrae** [GCTC-NTGC/gc-digital-talent#7393]
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/537a8a10-7667-4f82-9ad7-7b5a588378f8)

## ✨ Summary

In an attempt to alleviate some of the design, functional, and accessibility concerns around our existing `combobox`/`mutliselect` component, we've gone ahead and designed a new set of requirements and mockups that should be implemented in its stead. After collaborating with @esizer on the design, these changes might require moving to a different, more flexible plugin. Whatever we choose, it should be able to meet the requirements defined below. 

### Core experience

Generally speaking, this component should look and behave like a native select element when empty, with the addition of the search icon to the left, and the ability to type into the input to filter the list of options.

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/55e0f0ae-ec13-4a40-a89b-59263b2d4122)

When clicked, the results list should immediately appear, showing a count of the total available options and the total list of options for selection.

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/02ad64d1-87a3-4505-a47e-831e8afa4798)

When the user begins to type, a loading state should appear if the query takes longer than being near-instantaneous. Note that a `clear` button appears when a typed value is entered, allowing the user to quickly clear a search. Clicking the clear button should immediately refocus the input so that the user can continue typing their new search.

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/dad643b9-b35b-458d-9ba5-814cd389c3dd)

Results should be visually returned after each keystroke. Note however, that each keystroke should NOT trigger a new AT readout - ATs should only be informed of the results if the user uses the `enter` or `down` keys to trigger and focus the results list. Otherwise they'll be getting non-stop information as they type. Typing a search changes the label to indicate how many results match the total number of options, as well as adds a button to clear the search.

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/1e5195e5-b4d9-4cf9-95b3-3fc77bac86a8)

When an item is selected, it should be given a checkmark, and both the icon and text should change to `primary.darker`.

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/3f1bf33c-8430-4759-948d-e107b90f6f31)

#### Other core states

This input should adopt our common input states, including the optional, permanently-visible data context element and input-level error messages. Note that the actual results dropdown always overlaps objects in the DOM directly below. This includes the data context well, the error message well, and the multi-select selection summary outlined below.

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/6a2ec90e-20f3-4a99-8693-70c66ce2a459)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/e6abd9ef-d06e-41c8-a497-c0bd85e5471d)

#### Core keyboard behaviour

Focusing the input should appear like focusing the single-select variant. 

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/c705162a-2003-4323-afb7-44584604a0c6)

Initial keyboard interactions:
- The `down` arrow key should open the results list, read out the number of results, and focus the first result - note that the `clear search` button is actually at the **end** of the cycle, meaning the user can immediately hit the `up` arrow to target it
- If the user chooses to instead type a query into the input and hit `enter`, the loading state should show and be communicated, at which point the results list should replace it, read out the results count, and be focused for cycling 
- The user should be able to navigate the results list with the `up` and `down` arrow keys, cycling them around in a loop
- The user should be able to select items using `enter` (see variants for more details)
- The user should be able to collapse the results list using `esc`, returning focus to the input
- The user should be able to collapse the results list using `tab`, focusing the next interactive element (the clear button, if available)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/6114cab8-6b6b-4913-95b1-db737a10d17f)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/8a74512d-55af-484b-8a2e-4c4da1ce6404)

When focusing an option in the options list that is currently selected, it should appear as follows. Note the font color change for focused elements that are colored `primary.darker` - they should always be `black` on `focus`.

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/869aa1f3-8881-4909-8f72-77ab6d1a521a)

### Single-select variants

Single-select variants allow the user to select a single option from the list of options. Selecting an item from the list should:
- Replace the typed value with the selected value
- Change the selected item's state in the options list to indicate it is currently selected

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/4e00a24b-ea6c-4037-9298-6bdfd0409096)

There are a couple nuances to consider with this variant:
- We'll want to validate the input's contents to ensure that typed values match an option in the list (this accounts for the behaviour of a user typing an option exactly right and not selecting it from the list)
- Typing in the input after making a selection should allow the user to modify the value as regular characters (selecting an item and then clicking back into the input should allow you to continue typing or backspace the word)
- Altering the selection by typing should also ensure that "selected" items in the list have their states updated appropriately

#### Single-select keyboard navigation

- Selecting an item with `enter` should close the results list and focus the input

### Multi-select variants

Multi-select variants allow the user to select one or more items from the list.

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/f4a67a68-4bab-498a-bc37-ee81ca09f474)

When items are selected, the component changes to show a permanent list of selected items underneath the dropdown. The permanent list that appears contains additional labels and controls for clarity and ease of use:
- "# options selected" label
- "Clear all selections" button that removes all selected items and resets the component to its default state
- Selected items represented as a new chip-style element that allows the user to click the chip to remove the selected item
- When multiple items are selected, they should wrap within the container.

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/0017be25-402f-44f1-be0e-a225ea3b0c32)

#### Mutli-select keyboard navigation

If the component contains selections, hitting `tab` while focusing the input or the results list should: 
- Focus the "clear" button if present, otherwise focus the first selected item
- Allow the user to cycle through the list of selected items using the `up` and `down` arrow keys
  - Note that the first focused item in this list should be the selected items; using the `up` arrow key on the first item in the list will cycle the user to end of the loop, focusing the `clear all selections` button
- `tab` to the next focusable element on the page, skipping any remaining selections

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/c4b394a7-c2b8-49c1-891d-fe7236ae708c)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/6960386/23335d03-d0c7-43f8-aee4-37f36fd981fe)

**Comment by substrae** [comment]
Re-opening at @mnigh's request, given the previous issue is already tied to a PR.

**Comment by RM-1978** [comment]
@esizer and I took another look at this together, 

We feel that this issue is well-though out and  described! That said, we wondered about the following implementation tweaks: 
 - Should  we present a search icon (which has certain connotations for the user from other experiences) if the component behaves like a native select element and we use a search element elsewhere on our platform? We recommend presenting no icon at all in the  input text box to ensure that users do not mistake this component  for a search component.
 -  If the options list appears right away when a user activates what appears to be a text input element (either by clicking or tabbing into it), then will we in advertently create cognitive accessibility barriers for the user? In addition, presenting the list  necessitates that the user exit the component and come back to see what chips appear for their selections; or they scroll through the entire list to see what they have selected. We recommend not changing the default functionality of the combobox and instead present the user with dropdown options once they have begun typing or activated the chevron (by clicking it or pressing the down arrow key). This way,  the input component and the drop down list will not interfere with the users view of the chips.
 - Should we propagate this styling for the chips to other platform instances. 
 - We agree wholeheartedly with presenting visual users with the same information as non-visual users when the options are loading :) 
 - When a list item is selected, it is given a checkmark, and both the icon and text change to primary.darker, @esizer will make sure that the user can unselect the option by activating it a second time (Click or navigate using the keyboard and taps enter).

**Comment by tristan-orourke** [comment]
It's up to @substrae to make any final calls on changes, but I want to share my thoughts on @RM-1987's suggestions.

- It seems to me like this component _does_ behave like a search component, as well as like a select component, and that having the search icon as well as the "dropdown menu chevron" helps indicate this.
- Similarly, I think it would be important to indicate as early as possible that this is _not_ a simple text input element. However, I'm am also unsure how the dropdown selections interact with the chips. Does the dropdown menu cover the "current selection" chips"? 
- Great point, unselecting also needs to be an acceptance criteria
- Yeah, do we want to style other chips?
- :+1:

**Comment by substrae** [comment]
@RM-1978 @tristan-orourke 

- Tristan has it right - this is indeed a search input. People click into it and can type a query to filter a list of search results. The difference is that they can simply choose a result from the list without typing if they-so-choose.
- There's a lot to unpack here, and @esizer and I have discussed and settled on most of them in the past already, but a few things I want to make clear:
  - To be super clear, a combobox is a combination of a search input and a select input
  - As such, the combobox needs to incorporate the interaction signals provided by those two elements: a search icon and text field for typing a search query AND a visual dropdown list of results to select from
  - The selection list containing the chips appears underneath the dropdown, yes, because the alternatives presented stronger accessibility concerns such as the input changing location as selections are made, the dropdown list being pushed off screen, and more
    - The dropdown list shows which items are selected, and to a view a summary is one key (`esc`) away
- Yes
- No - these chips are unique the multi-select component. They have to occupy much less space and provide fewer controls in terms of component properties.
- Yes

@petertgiles @gobyrne 

I don't think this is going to work (see the comment above) - the existing chips are significantly larger than what I've provided here, so it's going to make the visual experience cluttered and harder to consume.

**Comment by petertgiles** [comment]
Our current combobox doesn't initialize with a value correctly.  If this fixes that issue then https://github.com/GCTC-NTGC/gc-digital-talent/issues/7770 can be closed.
